### PR TITLE
Crawling folders recursively

### PR DESCRIPTION
**Pull request by Antonhansel**

### CODE REVIEW

File: lib/update.js [hunk 1]
```diff
@@ -1,9 +1,79 @@
...
-  cb(null, new Date());
+
+  async.waterfall([
+    function refreshToken(cb) {
+      var apiUrl = "https://api.box.com/oauth2/";
+      var url = '/token';
+      request(apiUrl)
+      .post(url)
+      .set('Content-Type', 'application/x-www-form-urlencoded')
+      .send({grant_type: 'refresh_token',
```
Comment (Neamar) [inline lib/update.js]: ({\n

File: lib/update.js [hunk 2]
```diff
@@ -1,9 +1,79 @@
...
+        refresh_token: serviceData.refresh_token,
+        client_secret: config.box.secret,
+        client_id: config.box.api})
```
Comment (Neamar) [inline lib/update.js]: \n})

File: lib/update.js [hunk 3]
```diff
@@ -1,9 +1,79 @@
...
+      .end(cb);
+    },
+    function parseFolders(data, cb) {
+      // saving the new tokens
+      serviceData = data.body;
+      // 0 is the default id of the root folder
+      getSubFolder(serviceData, cursor, 0, cb);
+    },
+    function sendFiles(files, cb) {
+      console.log(files);
```
Comment (Neamar) [inline lib/update.js]: console ;)

(maybe next PR)

File: lib/update.js [hunk 4]
```diff
@@ -1,9 +1,79 @@
...
+      cb(null);
+    }
+    ], function(err) {
+      console.log('ERROR ____________' + err);
+      cb(null, new Date(), serviceData);
```
Comment (Neamar) [inline lib/update.js]: cb(err

File: lib/update.js [hunk 5]
```diff
@@ -1,9 +1,79 @@
...
+      .end(cb);
+    },
+    function processFolderContent(content, cb) {
+      var allFilesId = [];
+      var total = content.body.item_collection.total_count;
+      if(total === 0) {
+        cb(null, allFilesId);
+      }
+      var processed = 0;
+      content.body.item_collection.entries.forEach(function(item) {
```
Comment (Neamar) [inline lib/update.js]: async.each

File: lib/update.js [hunk 6]
```diff
@@ -1,9 +1,79 @@
...
+        if(item.type === 'folder') {
```
Comment (Neamar) [inline lib/update.js]: maybe you can filter before?

File: lib/crawler.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+var request = require('supertest');
+
+var Crawler = function Crawler() {
+  this.queues;
+  this.serviceData;
+  this.cursor;
+  this.taskQueue;
+};
+
+Crawler.prototype.processTask = function(folder, callback) {
```
Comment (Neamar) [inline lib/crawler.js]: // Explanation?

File: lib/crawler.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        .expect(200)
+        .end(cb);
+    },
+    function processFolderContent(content, cb) {
+      var allFilesId = content.body.item_collection.entries.filter(function(item) {
+        return item.type !== 'folder';
+      }).map(function(item) {
+        return item.id;
+      });
+      // push files in queues
```
Comment (Neamar) [inline lib/crawler.js]: in provider queue

File: lib/crawler.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      allFilesId.forEach(function(item) {
+        self.queues.addition.push(item);
+      });
+      var folders = content.body.item_collection.entries.filter(function(item) {
+        return item.type === 'folder';
+      });
+      self.taskQueue.push(folders);
```
Comment (Neamar) [inline lib/crawler.js]: // Push folder in recursive queue

File: lib/crawler.js (hunk 7 above)
Comment (Neamar) [inline lib/crawler.js]: Use either cb() or next()

File: lib/update.js [hunk 10]
```diff
@@ -1,9 +1,46 @@
...
+    function parseFolders(data, cb) {
+      // saving the new tokens
+      serviceData = data.body;
+      // init of the crawler
+      var crawler = new Crawler();
+      crawler.serviceData = serviceData;
+      crawler.queues = queues;
+      crawler.cursor = cursor;
+      // create the taskQueue for the crawler
+      crawler.taskQueue = async.queue(crawler.processTask.bind(crawler), 5);
```
Comment (Neamar) [inline lib/update.js]: why is this not done on Crawler init?

File: lib/update.js [hunk 11]
```diff
@@ -1,9 +1,46 @@
...
+      crawler.taskQueue.drain = cb;
```
Comment (Neamar) [inline lib/update.js]: idem, abstract this on Crawler and emit an event once drained?

File: lib/update.js [hunk 12]
```diff
@@ -1,9 +1,46 @@
...
+      // push the folders to the taskQueue
```
Comment (Neamar) [inline lib/update.js]: the main folder (id 0)

Comment (Neamar) [inline lib/update.js]: or root folder?

File: lib/crawler.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+var request = require('supertest');
+
+var Crawler = function Crawler(serviceData, queues, cursor, cb) {
+  this.queues = queues;
+  this.serviceData = serviceData;
+  this.cursor = cursor;
+  this.finalCb = cb;
+  this.taskQueue = async.queue(this.processTask.bind(this), 5);
+  this.taskQueue.drain = function() {
+    if(this.taskQueue.length() === 0) {
```
Comment (Neamar) [inline lib/crawler.js:17]: maybe explain why?

File: config/configuration.js [hunk 14]
```diff
@@ -35,8 +35,8 @@ module.exports = {
   appSecret: process.env.ANYFETCH_API_SECRET,
 
   box: {
-    api: process.env.BOX_API_ID,
-    secret: process.env.BOX_API_SECRET
+    api: process.env.BOX_API_ID || 'api_id',
```
Comment (Neamar) [inline config/configuration.js:38]: what for? :\ not a huge fan of fake credentials

File: lib/crawler.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+      // push files in provider queue
+      allFilesId.forEach(function(item) {
+        self.queues.addition.push(item);
+      });
+      var folders = content.body.item_collection.entries.filter(function(item) {
+        return item.type === 'folder';
+      });
+      // push folders in the recursive queue
+      if(folders.length > 0) {
+        self.taskQueue.push(folders);
```
Comment (Neamar) [inline lib/crawler.js:53]: Explain why the if()

File: lib/crawler.js [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+      }
+      cb(null);
+    }
+  ], cb);
```
Comment (Neamar) [inline lib/crawler.js]: maybe you should wrap the orignal cb() and display console.warn(err)

File: lib/helpers/upload.js [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    function getFileContent(cb) {
+      request(apiUrl)
+        .get(url)
+        .set('Content-Type', 'application/x-www-form-urlencoded')
+        .set('Authorization', 'Bearer ' + accessToken)
+        .expect(200)
+        .end(cb);
+    },
+    function processFolderContent(content, cb) {
+      console.log(content);
```
Comment (Neamar) [inline lib/helpers/upload.js:22]: console

File: lib/helpers/upload.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+      cb(null);
+    }
+    ], function(err) {
+      console.log('Error in upload:' + err);
+      cb(null);
+    });
+
+  // anyfetchClient.postDocument(formattedContact, rarity.slice(1, cb));
```
Comment (Neamar) [inline lib/helpers/upload.js:30]: next PR?

File: package.json [hunk 19]
```diff
@@ -1,36 +1,38 @@
 {
-  "name": "anyfetch-PROVIDER_NAME-PROVIDER",
+  "name": "box-provider.anyfetch.com",
   "version": "0.1.0",
-  "author": "You <you@example.com>",
+  "author": "Antonhansel antoninribeaud@gmail.com",
```
Comment (Neamar) [inline package.json]: `Antonhansel <antoninribeaud@gmail.com>`

File: package.json [hunk 20]
```diff
@@ -1,36 +1,38 @@
...
   "bin": {
     "server": "./bin/server"
   },
   "scripts": {
     "start": "node ./bin/server",
     "test": "NODE_ENV=test mocha -R spec -t 25000",
-    "lint": "jshint bin/ config/ lib/ test/ app.js && jscs bin/ config/ lib/ test/ app.js"
+    "lint": "jshint bin/ config/ lib/ app.js && jscs bin/ config/ lib/ app.js"
```
Comment (Neamar) [inline package.json]: why no test?

File: package.json [hunk 21]
```diff
@@ -1,36 +1,38 @@
...
   },
   "main": "./lib/",
   "repository": {
     "type": "git",
-    "url": "https://github.com/YOU/PROVIDER_NAME.provider.example.com"
+    "url": "https://github.com/anyfetch/box-provider.anyfetch.com"
```
Comment (Neamar) [inline package.json]: AnyFetch

File: config/configuration.js (hunk 14 above)
Comment (Antonhansel) [inline config/configuration.js:38]: So the tests can be run without having to define any env variables, should it be removed?

File: package.json [hunk 22]
```diff
@@ -1,36 +1,38 @@
...
   "keywords": [
     "anyfetch",
     "provider",
-    "PROVIDER"
+    "box"
   ],
   "dependencies": {
     "anyfetch-provider": "^3.1.0",
-    "async": "^0.9.0",
+    "async": "git://github.com/Antonhansel/async.git",
```
Comment (Neamar) [inline package.json]: grumpf, not a huge fan of this (slower deployments)

File: test/init.js [hunk 23]
```diff
@@ -1,15 +1,15 @@
...
 
-describe("GET /init/connect", function() {
-  it("should redirect to PROVIDER", function(done) {
-    request(app).get('/init/connect?code=123')
-      .expect(302)
-      .expect('Location', /localhost:8000/)
-      .end(done);
-  });
-});
+// describe("GET /init/connect", function() {
```
Comment (Neamar) [inline test/init.js]: why comment?

File: lib/update.js [hunk 24]
```diff
@@ -1,9 +1,39 @@
...
+var Crawler = require('./crawler.js');
+
+
 module.exports = function updateAccount(serviceData, cursor, queues, cb) {
-  // Update documents from provider
-  // You may define this as an helper function
-  // this function is pinged on update, with the data stored by retrieveTokens()
-  // You can do queues.worker_name.push(task)
-  cb(null, new Date());
+
```
Comment (Neamar) [inline lib/update.js]: \n

File: lib/helpers/upload.js (hunk 18 above)
Comment (Antonhansel) [inline lib/helpers/upload.js:30]: yes

File: lib/helpers/upload.js (hunk 17 above)
Comment (Antonhansel) [inline lib/helpers/upload.js:22]: Next PR

File: lib/helpers/upload.js [hunk 25]
```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+// var rarity = require('rarity');
+var request = require('supertest');
+var async = require('async');
+
+
+module.exports = function upload(file, anyfetchClient, accessToken, cb) {
+
```
Comment (Neamar) [inline lib/helpers/upload.js]: \n

File: lib/workers.js [hunk 26]
```diff
@@ -1,9 +1,18 @@
 'use strict';
 
+var log = require('anyfetch-provider').log;
+
+// var uploadContact = require('./helpers/upload.js');
```
Comment (Neamar) [inline lib/workers.js:5]: commented code

File: lib/workers.js [hunk 27]
```diff
@@ -1,9 +1,18 @@
...
+
 module.exports.addition = function additionQueueWorker(job, cb) {
   // Send data to AnyFetch.
   // You may define this as an helper function
   // Job is an item with your task, an anyfetch client, and data stored by retrieveTokens()
+  log.info({
+    name: 'addition',
+    identifier: job.task.identifier
+  }, "Uploading");
+  // uploadContact(job.task, job.anyfetchClient, job.serviceData.accessToken, cb);
```
Comment (Neamar) [inline lib/workers.js:15]: commented code

File: test/test-crawler.js [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+describe("Testing the crawler", function() {
+
+  var filesPushed = [];
+
+  var fakeQueue = {
+    addition: []
+  };
+
+  before(function(done) {
+
```
Comment (Neamar) [inline test/test-crawler.js:16]: \n

File: test/test-crawler.js [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+require('should');
+var nock = require('nock');
+
+var update = require('../lib/update.js');
+var config = require('../config/configuration.js');
+
+
+describe("Testing the crawler", function() {
+
```
Comment (Neamar) [inline test/test-crawler.js]: \n

File: test/test-crawler.js [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+    .post('/oauth2/token', "grant_type=refresh_token&refresh_token=fake_token&client_secret=" + config.box.secret + "&client_id=" + config.box.api)
+    .reply(200, {
+      access_token: 'fake_access'
+    });
+
+    var mockRootFolder =
+    nock('https://api.box.com:443')
+    .get('/2.0//folders/0')
+    .reply(200, {
+      "item_collection": {
```
Comment (Neamar) [inline test/test-crawler.js:26]: let's put that in a file, such as data/mock/2.0//folders/0/index.json

File: test/test-crawler.js [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+    var mockRefreshToken =
+    nock('https://api.box.com:443')
+    .post('/oauth2/token', "grant_type=refresh_token&refresh_token=fake_token&client_secret=" + config.box.secret + "&client_id=" + config.box.api)
+    .reply(200, {
+      access_token: 'fake_access'
+    });
+
+    var mockRootFolder =
+    nock('https://api.box.com:443')
+    .get('/2.0//folders/0')
```
Comment (Neamar) [inline test/test-crawler.js:24]: double slash?

File: test/test-crawler.js [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+          "id": "2",
+          "sha1": "134b65991ed521fcfe4724b7d814ab8ded5185dc",
+          }
+        ],
+      }
+    });
+
+    var mockSubFolder =
+    nock('https://api.box.com:443')
+    .get('/2.0//folders/1')
```
Comment (Neamar) [inline test/test-crawler.js:42]: //

File: test/test-crawler.js [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+    .reply(200, {
+      "item_collection": {
```
Comment (Neamar) [inline test/test-crawler.js:44]: idem, let's put that in a file

File: lib/workers.js (hunk 27 above)
Comment (Antonhansel) [inline lib/workers.js:15]: next PR

File: lib/workers.js (hunk 26 above)
Comment (Antonhansel) [inline lib/workers.js:5]: next PR

File: package.json (hunk 22 above)
Comment (Antonhansel) [inline package.json]: Is there another way to use our async fork? Is it really a lot slower?

File: config/configuration.js (hunk 14 above)
Comment (Neamar) [inline config/configuration.js:38]: No better idea for now :(

**Pull request merged**